espei/error_functions/modified_zpf_error.py

```diff
@@ -0,0 +1,116 @@
+                        ):
+  print('Running new zpf')
+  comps = ['CU', 'MG', 'VA']
+  models = instantiate_models(dbf, comps, phases, parameters=parameters)
+  cbs = build_callables(dbf, comps, phases, models=models, parameter_symbols=parameters.keys(), 
+                        additional_statevars={v.P, v.T, v.N}, build_gradients=True, build_hessians=True)
+  if parameters is None:
+    parameters = {}
+  prob_error = 0
+  phase_condition_data = {}
+  phase_chem_pot_data = {}
+  for phase in phases:
+    phase_condition_data[phase] = PhaseConditions(models, phase, np.array([1.0, 101325.0, 300.0]), cbs['GM']['callables'][phase],
+                                                                                                   cbs['GM']['grad_callables'][phase],
+                                                                                                   cbs['GM']['hess_callables'][phase],
+                                                                                                   parameters.values())
+  prob_error = 0.0
+  for data in zpf_data:
+    # unpack the data
+    phase_regions = data['phase_regions']
+    data_comps = data['data_comps']
+    if set(data_comps) != set(comps):
+      print('Warning: Only implemented for CU-MG')
+      return 0
+    weight = data['weight']
+    dataset_ref = data['dataset_reference']
+    for region, region_eq in phase_regions.items():
+      for current_statevars, comp_dicts in region_eq:
+        state_array = np.array([1.0, current_statevars[v.P], current_statevars[v.T]])
+        for phase in phases:
+          phase_condition_data[phase].set_state_vars(state_array)
+        sample_chemical_potentials = np.column_stack((np.linspace(-10000.0, 10000.0, chem_samples), np.zeros(chem_samples))).transpose()
+        for phase in phases:
+          phase_chem_pot_data[phase] = single_phase_opt_at_chem_pot(phase_condition_data[phase], sample_chemical_potentials)
+        chem_pot_errors = np.zeros(chem_samples)
+        chem_pot_minima = np.minimum.reduce(list(phase_chem_pot_data.values()))
+        for current_phase, cond_dict in zip(region, comp_dicts):
+          cond_dict, phase_flag = cond_dict
+          for key, val in cond_dict.items():
+            if key.__str__() != 'X_MG' and key.__str__() != 'X_CU':
+              print(key)
+              print('Not Implemented yet.')
+              continue
+            if val is None or val is np.nan:
+              chem_pot_errors += np.square((phase_chem_pot_data[current_phase][0,:] - chem_pot_minima[0,:]).transpose())
+            else:
+              # X_MG is listed second.
+              current_composition = np.array([1.0-val, val])
+              if key.__str__() == 'X_CU':
+                current_composition = np.array([val, 1.0-val])
+              # If the driving force is negative, set it to zero. This is a hack since our sampling may not be dense enough. Should not be necessary when Newton's method has been implemented.
```

Comment:
I'm not sure I understand why we are setting negative driving forces to zero. Conventionally, if a phase has a positive driving force, that means it is more stable (the energy is more negative) than the hyperplane it is being compared to. If the driving force is negative, that means that phase is less stable (and does not want to form).

Using the convention above with the ZPF data, we should be optimizing so that the phase of interest has a zero driving force and we need to destabilize it if the driving force is positive and stabilize it if the driving force is negative. In either case, the log probability we calculate from the normal distribution is should be continuous and at a minimum when the driving force is zero, so setting the driving forces to zero artificially doesn't make sense to me.

Reply:
The way we are calculating the loss forces all driving forces to be positive. This is because for each chemical potential we calculate a plane which lies below all of the phases (in fact this seems like it would be true for the original loss functions as well). The only reason some of the driving forces may be negative is because we aren't exactly solving the optimization problem for each hyperplane (at the moment we just take 1000 samples). If we replace this by a sampling and Newton's method, then the negative driving forces should go away.

Reply:
In the currently implemented loss function, it depends on the data. 

If there's only one phase composition in the tie-line region, that means the target hyperplane will only be described by the all phase hyperplane at that particular composition and therefore the single phase hyperplane must be equal to or above that hyperplane by construction. This is the case for virtually all Cu-Mg data, which you are probably looking at.

If you look at some data where there are multiple phase compositions are described, the target hyperplane is a mean of multiple mulit-phase hyperplanes at the different compositions. It's then possible that hyperplane for the single phase calculation is below the mean target hyperplane, even though it will never be below the multi-phase hyperplane at that composition.

--


For what's happening here, is it because the we are taking a `linspace` of chemical potentials? Would it solve the problem to use pycalphad's lower_convex_hull to calculate the chemical potentials on your Gibbs energy grid?

Reply:
The issue isn't the collection of chemical potentials, but rather the single_phase_opt_at_chem_pot, which samples some points to calculate the single phase minimum. Consequently, the single phase minimum is not calculated accurately enough for the LAVES_C15 phase. I am confident that this can be remedied by using a Newton's method to refine the best sample point.

I also think that if you have two hyperplanes which are lie below all of the phase energies, then their average will also lie below all of the phase energies, so even the regular ESPEI algorithm shouldn't produce a negative driving force.

Reply:
> The issue isn't the collection of chemical potentials, but rather the single_phase_opt_at_chem_pot, which samples some points to calculate the single phase minimum. Consequently, the single phase minimum is not calculated accurately enough for the LAVES_C15 phase. I am confident that this can be remedied by using a Newton's method to refine the best sample point.

Can you fix it by using the same grid of site fractions/energies for all the phases?

> I also think that if you have two hyperplanes which are lie below all of the phase energies, then their average will also lie below all of the phase energies, so even the regular ESPEI algorithm shouldn't produce a negative driving force

Yeah, you're right on this. Thanks!